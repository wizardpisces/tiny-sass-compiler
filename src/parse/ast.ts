export enum NodeTypes {
    TEXT = 'TEXT',  // TEXT = (TEXT\s+ | TEXT\s+)*
    VARIABLE = 'VARIABLE', // VARIABLE.value === variable's name , expression deleted after evaluation
    PUNC = 'PUNC',  // punctuation: parens((|)), comma(,), semicolon(;) etc.
    OPERATOR = 'OPERATOR',   // arithmeticOperator | comparisonOperator
    VAR_KEY = 'VAR_KEY', // to solve "TEXT-#{VARIABLE}" , expression replaced after evaluation
    PLACEHOLDER = 'PLACEHOLDER', // %TEXT
    KEYWORD = 'KEYWORD',
    IDENTIFIER = 'IDENTIFIER',
    EMPTY = 'EMPTY',
    UNKNOWN = 'UNKNOWN',
    SELECTOR = 'SELECTOR',
    /**
     * https://sass-lang.com/documentation/values/lists
     * any expressions separated with spaces or commas count as a list;
     * iterable eg: $each value in list
    */
    LIST = 'LIST',
    BINARY = 'BINARY',
    CALL = 'CALL',

    /**
     * Statement 
     * */
    BLOCK = 'BLOCK',
    BODY = 'BODY',
    RULE = 'RULE',
    DECLARATION = 'DECLARATION',

    // keyword statement
    PLUGIN = 'PLUGIN',
    IMPORT = 'IMPORT',
    INCLUDE = 'INCLUDE',// use mixin
    EXTEND = 'EXTEND',// combind repeated css
    MIXIN = 'MIXIN', // allow you to define styles that can be re-used throughout your RootNode.
    CONTENT = 'CONTENT',// must end with @return
    FUNCTION = 'FUNCTION',// must end with @return

    /**
     * internal Atrule
     */
    MediaFeature = 'MediaFeature',
    MediaPrelude = 'MediaPrelude',
    KeyframesPrelude = 'KeyframesPrelude',
    MediaQuery = 'MediaQuery',
    Atrule = 'Atrule',

    // exceptions
    ERROR = 'ERROR',

    // control flow
    RETURN = 'RETURN',// always return a text

    // choice statement
    IFSTATEMENT = 'IFSTATEMENT',

    //Loops
    EACHSTATEMENT = 'EACHSTATEMENT',

    RootNode = 'RootNode',

}
/**
 * new keyword add steps
 * 0. add test cases to test folder
 * 1. modify ast.ts (add new Node ast etc)
 * 2. modify lexical keywords
 * 4. 
 */

export type keywordType = '@extend'
    | '@mixin' | '@content' | '@include'
    | '@import'
    | '@if' | '@else'
    | '@error'
    | '@each'
    | '@function' | '@return'
    | '@plugin'

export type internalAtRuleNames = 'media' | 'keyframes' | 'support' | 'charset' | 'font-face' | string;

export type puncType = '(' | ')' | ',' | ';' | '#' | '{' | '}'
export type assignPuncType = ':'
export type arithmeticOperator = '+' | '-' | '*' | '/' | '%'
export type comparisonOperator = '>' | '<' | '>=' | '<=' | '==' | '!='

export const locStub: SourceLocation = {
    filename: '',
    start: { line: 1, column: 1, offset: 0 },
    end: { line: 1, column: 1, offset: 0 }
}

export interface Position {
    offset: number // from start of file
    line: number
    column: number
}

export interface SourceLocation {
    filename: string
    start: Position
    end: Position
    // source: string
}

export interface Node {
    [key: string]: any
    type: NodeTypes
    loc: SourceLocation
}

export interface TextNode extends Node {
    type: NodeTypes.TEXT
    value: string
}
export interface EmptyNode extends Node {
    type: NodeTypes.EMPTY
}

export interface VariableNode extends Node {
    type: NodeTypes.VARIABLE
    value: string
}
export interface IdentifierNode extends Node {
    type: NodeTypes.IDENTIFIER
    value: string
}
export interface PlaceholderNode extends Node {
    type: NodeTypes.PLACEHOLDER
    value: string
}

export interface VarKeyNode extends Node {
    type: NodeTypes.VAR_KEY
    value: string
}

export interface PuncNode extends Node {
    type: NodeTypes.PUNC
    value: puncType
}

export interface OperatorNode extends Node {
    type: NodeTypes.OPERATOR
    value: arithmeticOperator | comparisonOperator
}

export interface SelectorNode extends Node {
    type: NodeTypes.SELECTOR
    meta: MediaStatement['prelude'][] // helper to trace nested media parent
    value: TextNode | PlaceholderNode | ListNode | EmptyNode
}

// combined node
export interface BinaryNode extends Node {
    type: NodeTypes.BINARY
    operator: OperatorNode
    left: TextNode | VariableNode | BinaryNode
    right: TextNode | VariableNode | BinaryNode
}

export interface ListNode extends Node {
    type: NodeTypes.LIST
    value: SimpleExpressionNode[]
}

export interface CallExpression extends Node {
    type: NodeTypes.CALL
    id: IdentifierNode
    args: ArgsType
}

export type SimpleExpressionNode = TextNode | PuncNode | OperatorNode | VariableNode | VarKeyNode | BinaryNode | ListNode | CallExpression | EmptyNode

/* Statement */

export type Statement =
    BodyStatement
    | RuleStatement
    | DeclarationStatement
    | ImportStatement
    | IncludeStatement
    | ExtendStatement
    | MixinStatement
    | FunctionStatement
    | ErrorStatement
    | IfStatement
    | EachStatement
    | ReturnStatement
    | Atrule
    | PluginStatement

export type ArgsType = (TextNode | VariableNode | BinaryNode | DeclarationStatement)[]

/**
 * used with self designed statement eg: @function @mixin
 * will be deleted after evaluationï¼Œ so it will not exist in codegen step
 */
export interface BodyStatement extends Node {
    type: NodeTypes.BODY
    children: RuleStatement['children']
}

export interface RuleStatement extends Node {
    type: NodeTypes.RULE
    selector: SelectorNode
    children: (Statement | CodegenNode)[]
}

export interface DeclarationStatement extends Node {
    type: NodeTypes.DECLARATION
    left: VariableNode | TextNode | VarKeyNode
    right: ListNode | TextNode // ListNode before transform , TextNode after transform
}

// keyword statement
export interface ImportStatement extends Node {
    type: NodeTypes.IMPORT
    params: TextNode[]
}
export interface PluginStatement extends Node {
    type: NodeTypes.PLUGIN
    value: TextNode
}

export interface ContentPlaceholder extends Node { // @content
    type: NodeTypes.CONTENT
}

export interface IncludeStatement extends Node {
    type: NodeTypes.INCLUDE
    id: IdentifierNode
    args: ArgsType
    content?: BodyStatement
}

export interface ExtendStatement extends Node {
    type: NodeTypes.EXTEND
    param: TextNode | PlaceholderNode
}

export interface MixinStatement extends Node {
    type: NodeTypes.MIXIN
    id: IdentifierNode
    params: (VariableNode | DeclarationStatement)[]
    body: BodyStatement
}

export interface FunctionStatement extends Node {
    type: NodeTypes.FUNCTION
    id: IdentifierNode
    params: MixinStatement['params']
    body: BodyStatement
}

export interface ReturnStatement extends Node {
    type: NodeTypes.RETURN
    argument: DeclarationStatement['right']
}

export interface ErrorStatement extends Node {
    type: NodeTypes.ERROR
    value: ListNode
}

// choice statement
export interface IfStatement extends Node {
    type: NodeTypes.IFSTATEMENT
    test: BinaryNode | TextNode
    consequent: BodyStatement
    alternate: IfStatement | BodyStatement | null
}

export interface EachStatement extends Node {
    type: NodeTypes.EACHSTATEMENT
    left: VariableNode
    right: VariableNode
    body: RuleStatement
}


/**
 * interface prototype for css internal Atrule eg: @media @keyframes @font-face @charset etc
 * Todos: change @import by Atrule
 */
export interface Atrule extends Node {
    type: NodeTypes.Atrule
    name: internalAtRuleNames
    block: BodyStatement
    prelude: MediaPrelude | KeyframesPrelude
}
/**
 * @media ast tree (simplified) start  
 * (reference: https://github.com/csstree/csstree)
 * eg: When the width is between 600px and 900px OR above 1100px - change the appearance of <div>
 * @media screen and (max-width: 900px) and (min-width: 600px), (min-width: 1100px) {}
 * 
 */
export interface MediaFeature extends Node { // (min-width: 600px)
    type: NodeTypes.MediaFeature
    name: string
    value: SimpleExpressionNode // after transform , in codegen value will be TextNode
}
export interface MediaQuery extends Node {
    type: NodeTypes.MediaQuery
    children: (TextNode | MediaFeature)[] // TextNode contains special identifier: screen | and
}

export interface MediaPrelude extends Node {
    type: NodeTypes.MediaPrelude
    children: MediaQuery[] // splited by ','
}

export interface MediaStatement extends Atrule {
    name: 'media'
    prelude: MediaPrelude
}

export interface KeyframesPrelude extends Node {
    type: NodeTypes.KeyframesPrelude
    children: (TextNode | VarKeyNode)[]
}

export interface Keyframes extends Atrule {
    name: string // keyframes | -webkit-keyframes etc
    prelude: KeyframesPrelude
}

/**
 * @media ast tree end
 */

/* codeGenNode means ast tree that is transformed  */

export type CodegenNode = TextNode | ProgCodeGenNode | RootNode

export type ProgCodeGenNode = RuleStatement | EmptyNode | SelectorNode | DeclarationStatement | Atrule | MediaStatement | MediaPrelude | Keyframes | KeyframesPrelude

export type FileSourceMap = {
    [key: string]: string
}
export interface RootNode extends Node {
    type: NodeTypes.RootNode

    // codeGenNode ,use it to generate source map between files
    fileSourceMap: FileSourceMap

    children: (Statement | CodegenNode)[]
}

// work before ast transform
export function createKeyframesPrelude(children: KeyframesPrelude['children']): KeyframesPrelude {
    return {
        type: NodeTypes.KeyframesPrelude,
        children,
        loc: {
            start: children[0].loc.start,
            end: children[children.length - 1].loc.end,
            filename: children[0].loc.filename
        }
    }
}

export function createKeyframes(name: string, prelude: Keyframes['prelude'], block: Keyframes['block']): Keyframes {
    return {
        type: NodeTypes.Atrule,
        name,
        block,
        prelude,
        loc: {
            start: prelude.loc.start,
            end: block.loc.end,
            filename: prelude.loc.filename
        }
    }
}

export function createMediaFeature(declaration: DeclarationStatement): MediaFeature {
    return {
        type: NodeTypes.MediaFeature,
        name: declaration.left.value,
        value: (declaration.right as ListNode).value[0],
        loc: declaration.loc
    }
}

export function createMediaQuery(children: MediaQuery['children']): MediaQuery {
    return {
        type: NodeTypes.MediaQuery,
        children,
        loc: locStub
    }
}

export function createMediaPrelude(children: MediaPrelude['children']): MediaPrelude {
    return {
        type: NodeTypes.MediaPrelude,
        children,
        loc: locStub
    }
}

export function createMediaStatement(prelude: MediaStatement['prelude'], block: MediaStatement['block']): MediaStatement {
    return {
        type: NodeTypes.Atrule,
        name: 'media',
        block,
        prelude,
        loc: {
            start: prelude.loc.start,
            end: block.loc.end,
            filename: prelude.loc.filename
        }
    }
}

export function createIdentifierNode(id: TextNode): IdentifierNode {
    return {
        loc: id.loc,
        value: id.value,
        type: NodeTypes.IDENTIFIER
    }
}

export function createCallExpression(id: CallExpression['id'], args: CallExpression['args']): CallExpression {
    return {
        type: NodeTypes.CALL,
        id,
        args,
        loc: {
            start: id.loc.start,
            end: args.length ? args[args.length - 1].loc.end : id.loc.end,
            filename: id.loc.filename
        }
    }
}

export function createDeclarationStatement(left: DeclarationStatement['left'], right: DeclarationStatement['right']): DeclarationStatement {
    return {
        type: NodeTypes.DECLARATION,
        left: left,
        right: right,
        loc: {
            start: left.loc.start,
            end: right.length ? right[right.length - 1].loc.end : right.loc.end,
            filename: left.loc.filename
        }
    }
}

export function createListNode(list: SimpleExpressionNode[]): ListNode {
    return {
        type: NodeTypes.LIST,
        value: list,
        loc: {
            start: list[0].loc.start,
            end: list[list.length - 1].loc.end,
            filename: list[0].loc.filename
        }
    }
}

export function createMixinStatement(id: MixinStatement['id'], params: MixinStatement['params'], body: MixinStatement['body']): MixinStatement {
    return {
        type: NodeTypes.MIXIN,
        id,
        params,
        body,
        loc: locStub
    }
}

export function createFunctionStatement(id: FunctionStatement['id'], params: FunctionStatement['params'], body: FunctionStatement['body']): FunctionStatement {
    return {
        type: NodeTypes.FUNCTION,
        id,
        params,
        // always contain a ReturnStatement in the last line
        body,
        loc: locStub
    }
}

export function createReturnStatement(list: SimpleExpressionNode[]): ReturnStatement {
    return {
        type: NodeTypes.RETURN,
        argument: createListNode(list),
        loc: locStub
    }
}

export function createRuleStatement(selector: RuleStatement['selector'], children: RuleStatement['children'], loc: SourceLocation = locStub): RuleStatement {
    return {
        type: NodeTypes.RULE,
        selector,
        children,
        loc
    }
}

export function createContentPlaceholder(loc: ContentPlaceholder['loc'] = locStub): ContentPlaceholder {
    return {
        type: NodeTypes.CONTENT,
        loc
    }
}

export function createIncludeStatement(id: IncludeStatement['id'], args: IncludeStatement['args'], content?: IncludeStatement['content']): IncludeStatement {
    let node: IncludeStatement = {
        type: NodeTypes.INCLUDE,
        id,
        args,
        loc: locStub
    }

    if (content) {
        node.content = content
    }

    return node;
}

export function createIfStatement(test: IfStatement['test'], consequent: IfStatement['consequent'], alternate: IfStatement['alternate']): IfStatement {
    return {
        type: NodeTypes.IFSTATEMENT,
        test,
        consequent,
        alternate,
        loc: locStub
    }
}
export function createEachStatement(left: EachStatement['left'], right: EachStatement['right'], body: EachStatement['body']): EachStatement {
    return {
        type: NodeTypes.EACHSTATEMENT,
        left,
        right,
        body,
        loc: locStub
    }
}

export function createVarKeyExpression(value: VarKeyNode['value'], loc: Node['loc']): VarKeyNode {
    return {
        type: NodeTypes.VAR_KEY,
        value,
        loc
    }
}
export function createTextNode(value: TextNode['value'] = '', loc: Node['loc'] = locStub): TextNode {
    return {
        type: NodeTypes.TEXT,
        value,
        loc
    }
}

export function createSelectorNode(value: SelectorNode['value'] = createTextNode(), meta: SelectorNode['meta'] = []): SelectorNode {
    return {
        type: NodeTypes.SELECTOR,
        meta,
        loc: value.loc,
        value: value
    }
}

export function createEmptyNode(loc: Node['loc'] = locStub): EmptyNode {
    return {
        type: NodeTypes.EMPTY,
        loc
    }
}

export function createBodyStatement(children: BodyStatement['children']): BodyStatement {
    return {
        type: NodeTypes.BODY,
        children,
        loc: locStub
    };
}

export function createRootNode(children: RootNode['children'], fileSourceMap: RootNode['fileSourceMap'], loc: Node['loc'] = locStub): RootNode {
    return {
        type: NodeTypes.RootNode,
        children,
        fileSourceMap,
        loc
    }
}

export function createRuleFromBody(body: BodyStatement): RuleStatement {
    return createRuleStatement(
        createSelectorNode(createTextNode('')),
        body.children
    );
}

export function createRuleFromMedia(media: MediaStatement): RuleStatement {
    return createRuleStatement(
        createSelectorNode(
            createTextNode(''),
            [media.prelude]
        ),
        media.block.children
    );
}

export function createMediaFromRule(rules: RuleStatement[] | RuleStatement): MediaStatement {
    if (!rules.length) {
        rules = [rules as RuleStatement]
    }
    /**
     * merge media prelude list as one glued by 'and' idetifier 
     * */
    function mergeMediaPreludeList(preludeList: MediaStatement['prelude'][]): MediaStatement['prelude'] {

        function mergeMediaPreludeChildren(children: MediaPrelude['children']): MediaQuery {
            let mediaQueryChildren = children.reduce(
                (mediaQueryList: MediaQuery['children'], mediaQuery: MediaQuery) =>
                    mediaQueryList.concat(mediaQuery['children'])
                , [])

            return createMediaQuery(mediaQueryChildren)
        }

        let children: MediaPrelude['children'] = [];

        preludeList.forEach((prelude: MediaStatement['prelude']) => {
            if (children.length > 0) { // push 'and' identifer to glue mediaQuery
                children.push(createMediaQuery([createTextNode('and')]))
            }
            children = children.concat(prelude.children)
        })

        // return mediaPrelude with only one child
        return createMediaPrelude(children.length > 1 ? [mergeMediaPreludeChildren(children)] : children)
    }

    let prelude = mergeMediaPreludeList(
        Array.from(
            new Set(
                rules.reduce(
                    (preludeList: MediaStatement['prelude'][], rule: RuleStatement) =>
                        preludeList.concat(rule.selector.meta)
                    , [])
            )
        )
    )
    // reset rule selector meta
    rules.forEach((rule: RuleStatement) => rule.selector.meta = [])

    return createMediaStatement(prelude, createBodyStatement(rules as RuleStatement[]))
}