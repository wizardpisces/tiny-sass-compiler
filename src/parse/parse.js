import {
    defaultOnError,
    ErrorCodes
} from './errors';

import {
    NodeTypes
} from './ast';

import {
    debug,
    PRECEDENCE,
    fillWhitespace
} from './util'

/**
 * 
 * @param {lex processed stream method} input
 */

export default function parse(input) {

    let assign_right_end_condition = default_right_end_condition;

    function injectPosition(parseFn){
        return function(...args){
            input.eliminateWhitespace()
            let start = input.getCoordination(),
                ast = parseFn.apply(null, args),
                end = input.getCoordination();

            /**
             * adjust start position
             * if arguments exist it might be an evaluated expression eg: parse_assign left param
             */
            if (args && args.length > 0) {
                if (typeof args[0].start === undefined) {
                    input.croak(`[${parseFn.name}]: expect expression with start property!`)
                }
                start = start > args[0].start ? args[0].start : start;
            }
            return {
                loc:{
                    start,
                    end,
                },
                ...ast
            };
        }
    }

    let parserNamespace = {
        parse_assign: injectPosition(function parse_assign(left) {

             input.next();

             function parse_assign_right() {
                 let right = []
                 while (!assign_right_end_condition()) {
                     if (debug()) break;
                     let result = maybe_binary(parserNamespace.parse_atom(), 0);
                     right.push(result)
                 }
                 return right;
             }
             return {
                 type: NodeTypes.ASSIGN,
                 left: left,
                 right: {
                     type: NodeTypes.LIST,
                     value: parse_assign_right()
                 }
             }
        }),

        parse_atom: injectPosition(function parse_atom() {

            if (is_kw('@extend')) {
                input.next()
                return parse_extend();
            }

            if (is_kw('@mixin')) {
                input.next()
                return parse_mixin();
            }

            if (is_kw('@include')) {
                input.next()
                return parse_include();
            }

            if (is_kw('@import')) {
                input.next()
                return parse_import();
            }

            if (is_kw('@if')) {
                input.next()
                return parse_if();
            }

            if (is_kw('@each')) {
                input.next()
                return parse_each();
            }

            if (is_kw('@error')) {
                input.next()
                return parse_error();
            }

            let tok = input.peek();
            if (tok.type === NodeTypes.VARIABLE || tok.type === NodeTypes.PLACEHOLDER) {
                return input.next();
            }

            if (tok.type === NodeTypes.PUNC) {
                if (tok.value === "#") {
                    return maybe_key_var_wrapper(() => input.next())
                }
                return input.next()
            }

            if (tok.type === NodeTypes.TEXT) {
                return parse_consecutive_str()
            }

            defaultOnError({
                code: ErrorCodes.UNKNONWN_TOKEN_TYPE,
                loc: input.getCoordination(),
                tok
            })
        }),

        maybe_call: injectPosition(function maybe_call(exp) { //to resolve rotate(30deg) or url("/images/mail.svg") this kind of inner call expression
            let expr = exp();

            if (is_punc('(')) {
                return {
                    type: expr.type,
                    value: expr.value + '(' + delimited('(', ')', ',', parse_expression).map(expr => expr.value).join(',') + ')'
                }
            }
            return expr;
        })

    }
/**
 * end with ';' , eg:
 * 
 * $boder : 1px solid red;
 * 
 * end with ',' | ')' , eg:
 * 
 * @mixin test($param1:1,$param2:2){} // assign expression in @mixin or $include
  */

    function default_right_end_condition(){
        return is_punc(';')
    }

    function set_call_params_args_assign_right_end_condition(){
        assign_right_end_condition = () => is_punc(',') || is_punc(')');
    }

    function reset_assign_right_end_condition(){
        assign_right_end_condition = default_right_end_condition
    }

    function is_punc(ch) {
        let tok = input.peek();
        return tok && tok.type == NodeTypes.PUNC && (!ch || tok.value == ch) && tok;
    }

    function is_kw(kw) {
        let tok = input.peek();
        return tok && tok.type == NodeTypes.KEYWORD && (!kw || tok.value == kw) && tok;
    }

    function is_op(op) {
        let tok = input.peek();
        return tok && tok.type == NodeTypes.OPERATOR && (!op || tok.value == op) && tok;
    }

    function is_assign() {
        let tok = input.peek()
        return tok && tok.type === NodeTypes.ASSIGN;
    }

    function skip_punc(ch) {
        if (is_punc(ch)) input.next();
        // Todos: no mandatory skip for now
        // else {
        //     input.croak("Expecting punctuation: \"" + ch + "\"");
        // }
    }

    function delimited(start, stop, separator, parser) {// FIFO
        let a = [], first = true;
        skip_punc(start);

        while (!input.eof()) {
            if (debug()) break;
            if (is_punc(stop)) break;
            if (first) first = false; else skip_punc(separator);
            if (is_punc(stop)) break;

            a.push(parser());
        }
        skip_punc(stop);
        return a;
    }

    function maybe_binary(left, left_prec) {
        let tok = is_op()
        if (tok) {
            if (PRECEDENCE[tok.value] > left_prec) {
                input.next();//skip op
                return maybe_binary({
                    type: NodeTypes.BINARY,
                    operator: tok.value,
                    left: left,
                    right: maybe_binary(parserNamespace.parse_atom(), PRECEDENCE[tok.value])
                }, left_prec)
            }
        }
        return left;
    }

    function parse_child(selector) {
        let children = delimited("{", "}", ";", parse_expression);
        // if (children.length == 0) return FALSE;
        // if (block.length == 1) return block[0];
        return { type: NodeTypes.CHILD, selector, children };
    }

    function maybe_assign(exp) {
        let expr = exp();

        if (is_assign()) {
            return parserNamespace.parse_assign(expr)
        }

        /**
         * handle selector may contain key_var, which should be resolved to list ast
         * .icon-#{$size} {}
        */

        if(is_punc('#')){
            return maybe_assign(()=>{
                return {
                    type:NodeTypes.LIST,
                    value: expr.type === NodeTypes.LIST ? expr.value.concat(parserNamespace.parse_atom()) : [expr].concat(parserNamespace.parse_atom())
                }
            })
        }

        if (is_punc('{')) {
            return parse_child(expr) //passin selector
        }

        return expr;
    }

    function parse_extend() {
        return {
            type: NodeTypes.EXTEND,
            param: input.next()
        }
    }

    function parse_block_statement() {
        let children = delimited("{", "}", ";", parse_expression);
        return { type: NodeTypes.BODY, children };
    }

    /**
     * Todos: add @content kw and include body
     */
    function parse_mixin() {
        let id = {
            type: NodeTypes.IDENTIFIER,
            name: input.next().value
        },
        params = [];

        
        if(!is_punc('{')){
            /**
             * Support default params, which contains assign ':' symbol; which will be processed in parse_assign
             * @mixin replace-text($image,$x:default1, $y:default2) {
              */
            set_call_params_args_assign_right_end_condition()

            params = delimited('(', ')', ',', parse_expression);
        }

        reset_assign_right_end_condition()

        return {
            type: NodeTypes.MIXIN,
            id,
            params, // %extend like expr or func expr
            body: parse_block_statement()
        }
    }

    function parse_include() {
        let id = {
            type: NodeTypes.IDENTIFIER,
            name: input.next().value
        },
        args = [];// @include mixin1;

        if(!is_punc(';')){ 
            set_call_params_args_assign_right_end_condition()
            args = delimited('(', ')', ',', parse_expression); // extend like expr or call expr
        }

        reset_assign_right_end_condition()

        return {
            type: NodeTypes.INCLUDE,
            id,
            args,
        }
    }

    function parse_import() {

        function processFilenameExp(exp) {
            exp.value = exp.value.match(/^['"](.+)['"]$/)[1]
            return exp;
        }

        let delimitor = input.peek(),
            params = [];

        if (!delimitor.value.startsWith("'") && !delimitor.value.startsWith('"') ){
            input.croak(`@import expected with ' or " but encountered ${delimitor}`)
        }

        while(!is_punc(';')){  // @import 'foundation/code', 'foundation/lists';
            params.push( processFilenameExp(input.next()) );
            if(is_punc(',')){
                skip_punc(',')
            }
        }

        return {
            type: NodeTypes.IMPORT,
            params
        }
    }

    function parse_if(){
        let alternate = null,
            testExpression = true,
            blockStatement = [];

        testExpression = maybe_binary(parserNamespace.parse_atom(),0);

        if (!is_punc('{')){
            input.croak(`@if expect '{' but encountered '${input.next().value}'`)            
        }

        blockStatement = parse_block_statement();

        if (is_kw('@else')){
            input.next();
            let predictToken = input.peek();
            /**
             * check if it's a @else if  statement
              */
            if (predictToken.type === NodeTypes.TEXT && predictToken.value === 'if'){
                input.next();
                alternate = parse_if();
            }else{
                alternate = parse_block_statement()
            }
        }

        return { type: NodeTypes.IFSTATEMENT, test: testExpression, consequent: blockStatement, alternate: alternate }
    }

    function parse_each(){
        let left = parserNamespace.parse_atom();

        /**
         * skip "in" expression
          */

        input.next();

        let right = parserNamespace.parse_atom();

        skip_punc('{')
        let blockStatement = parse_expression()
        skip_punc('}')
        return {
            type: NodeTypes.EACHSTATEMENT,
            left,
            right,
            body: blockStatement
        }
    }

    function parse_list(endCheck = () => !default_right_end_condition()) {
        let list = []
        while (endCheck()) {
            list.push(parserNamespace.parse_atom())
        }
        return {
            type:NodeTypes.LIST,
            value: list
        }
    }

    function parse_error(){
        return {
            type:NodeTypes.ERROR,
            value: parse_list()
        }
    }
    
    

    function parse_consecutive_str() { //to resolve test skew(20deg) rotate(20deg);


        function read_while(predicate) {
            let tokens = [];

            while (!input.eof() && predicate(input.peek()))
                tokens.push(parserNamespace.maybe_call(() => input.next()));
            return tokens;
        }

        let tokens = read_while(tok => tok.type === NodeTypes.TEXT)

        return {
            type: NodeTypes.TEXT,
            value: fillWhitespace(tokens).map(tok => tok.value).join('')
        }
    }

    function parse_key_var_wrapper(){
        skip_punc('{')
        let var_key = input.next();
        if(var_key.type!==NodeTypes.VARIABLE){
            input.croak(`${var_key} should be a variable which starts with '$'`)
        }
        skip_punc('}')
        return {
            type:NodeTypes.VAR_KEY,
            value:var_key.value
        };
    }

    /**
     * 
     * #{var} 
     */

    function maybe_key_var_wrapper(exp){
        let expr = exp();

        if(is_punc('{')){
            return parse_key_var_wrapper()
        }

        expr.type = NodeTypes.TEXT;

        /**
         * color: #1212; or #selector{}
          */
        let nextToken = parserNamespace.parse_atom();

        if(nextToken.type!==NodeTypes.TEXT){
            input.croak(`[maybe_key_var_wrapper]: expect str token but received ${nextToken.value}`)
        }

        expr.value = expr.value + nextToken.value;

        return expr;
    }

    function parse_expression() {
        return maybe_assign(function () {
            return parserNamespace.parse_atom()
        })
    }

    function parse_prog(){
        let children = [];
        while (!input.eof()) {
            let result = parse_expression()
            children.push(result);
            if (!input.eof()) skip_punc(";");
        }
        return {
            type: NodeTypes.PROGRAM, 
            children
        };
    }

    function parse_toplevel() {
        return parse_prog()
    }

    return parse_toplevel()
}