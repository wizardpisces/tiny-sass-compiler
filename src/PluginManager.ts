import { CodegenNode, NodeTypes, RootNode, ProgCodeGenNode } from './parse/ast'

export type PluginContext = {
    remove(node: CodegenNode): void
    insert(node: CodegenNode): void
}

export type PluginFn = (node: CodegenNode, context: PluginContext) => void;

export type PluginObject = {
    enter?: PluginFn,
    leave?: PluginFn
}

export type Plugin = PluginFn | PluginObject

class PluginManager {
    _enterWalkers: Set<PluginFn> 
    _leaveWalkers: Set<PluginFn>
    _applied:boolean
    constructor() {
        this._enterWalkers = new Set()
        this._leaveWalkers = new Set()
        this._applied = false
    }

    _addEnterWalker(walker: PluginFn) {
        this._enterWalkers.add(walker)
    }

    _addLeaveWalker(walker: PluginFn) {
        this._enterWalkers.add(walker)
    }

    _walk(node: CodegenNode, list: CodegenNode[]) {
        this._runEnterWalkers(node, list)
        if (node.type === NodeTypes.RULE) {
            (node.children as CodegenNode[]).forEach((CHILDNode: CodegenNode, index: number) => {
                this._walk(CHILDNode, node.children as CodegenNode[])
            })
        }
        this._runLeaveWakers(node,list)
    }

    applyPlugins(ast:RootNode){
        if (this._applied){
            return;
        }else{
            this._applied = true;
        }

        (ast.children as ProgCodeGenNode[]).forEach((node: ProgCodeGenNode) => this._walk(node, ast.children as ProgCodeGenNode[]));
    }

    registerPlugin(plugin: Plugin) {
        if (typeof plugin === 'function') {
            return this._addEnterWalker(plugin)
        } else {
            plugin.enter && this._addEnterWalker(plugin.enter)
            plugin.leave && this._addLeaveWalker(plugin.leave)
        }
    }

    _createContext(visitedNode:CodegenNode, list: CodegenNode[]): PluginContext {
        return {
            remove(node: CodegenNode) {
                list.splice(list.indexOf(node),1)
            },
            insert(node: CodegenNode) {
                list.splice(list.indexOf(visitedNode) + 1, 0, node)
            }
        }
    }

    _runEnterWalkers(node: CodegenNode, list: CodegenNode[]) {
        this._enterWalkers.forEach(walker => walker.call(this, node, this._createContext(node, list)))
    }
    _runLeaveWakers(node: CodegenNode, list: CodegenNode[]) {
        this._leaveWalkers.forEach(walker => walker.call(this, node, this._createContext(node, list)))
    }
}

export const pluginManager = new PluginManager();

export function registerPlugin(plugin:Plugin){
    return pluginManager.registerPlugin(plugin)
}
export function applyPlugins(ast:RootNode){
    return pluginManager.applyPlugins(ast)
}