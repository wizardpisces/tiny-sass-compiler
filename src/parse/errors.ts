import { Position} from './ast'

export interface CompilerError extends SyntaxError {
    [key:string]:any
    code: ErrorCodes
    position?: Position
}

export function defaultOnError(error: CompilerError) {
    throw error
}

export const enum ErrorCodes {
    //parse error
    UNKNONWN_TOKEN_TYPE= 'UNKNONWN_TOKEN_TYPE'
}