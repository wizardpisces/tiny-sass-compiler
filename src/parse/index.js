import input_stream  from './input_stream.js'
import lexical from './lexical.js'
import parse from './parse.js'
import schema_check from './schema_check'

export default (scss) => schema_check(parse(lexical(input_stream(scss))))