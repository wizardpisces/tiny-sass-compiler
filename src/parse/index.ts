import input_stream from './input_stream'
import lexical from './lexical'
import parse from './parse'
import { RootNode } from './ast'
import { ParserOptions } from '@/options'
// import runtime_schema_check from './runtime_ast_check'

export default function baseParse(
    scss: string,
    options: ParserOptions
): RootNode {
    return parse(lexical(input_stream(scss)), options)
}