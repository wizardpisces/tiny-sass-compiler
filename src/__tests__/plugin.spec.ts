import { parse, transform, registerPlugin, generate } from '../index'
import { CodegenNode, NodeTypes } from '../parse/ast'
import { PluginContext } from '../PluginManager'
describe('plugin: tests', () => {
    const source = `
$stack:    Helvetica, sans-serif;
$primary: #333;

body .test{
  font: 100% $stack;
  color: $primary;
}

`.trim()
let result = 
`body .test{
  color:#333;
}
`
    function testPlugin(node: CodegenNode, context: PluginContext) {
        if (node.type === NodeTypes.ASSIGN && node.left.value === 'font') {
            context.remove(node)
        }
    }

    test('registerPlugin', () => {
        let parsedAst = parse(source, {
            filename: `default.scss`,
            source
        })
        transform(parsedAst, {
            filename: `default.scss`,
            source
        })
        registerPlugin(testPlugin)

        const { code } = generate(parsedAst)
        expect(code).toEqual(result)
        expect(code).toMatchSnapshot()
    })
})
