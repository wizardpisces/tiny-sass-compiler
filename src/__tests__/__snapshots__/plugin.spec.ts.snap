// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin: tests registerPlugin 1`] = `
"body .test{
  color:#333;
}
"
`;
